### PR TITLE
COMPRESS-342 random access of 7z files

### PR DESCRIPTION
**Pull request by PeterAlfreadLee**
Expose random access for 7z files in SevenZFile.
As files are compressed in `folder` in 7z specification, a `folder` may contain many files. A access to file N in a `folder` need to skip the previous 1...(N-1) files as they are compressed together in a folder.
This may result in a performance problem.

### CODE REVIEW

**Comment by bodewig** [comment]
Many thanks @PeterAlfreadLee 

I need to double-check but I think the sequential reading relies on the current file pointer, at least for entries that use streams inside of the same `Folder`. In that case mixing random and sequential access will not work and we need to make prominent notes in several places.

**Comment by bodewig** [comment]
Hmm `buildDecodingStream` does a lot of the heavy lifiting you've been doing yourself.

You could make your code work together with sequential access if you'd set `currentFolderIndex` and `currentFolderInputStream` as well, or we may find a way to parameterize `buildDecodingStream` and the new method would more or less boil down to calling the augmented `buildDecodingStream` followed by `getCurrentStream` (or rather a variant of it which wouldn't force you to set `currentEntryIndex` as well).

**Comment by PeterAlfreadLee** [comment]
@bodewig  Sure. Thank you so much.
I'll look into it and push my code today.

**Comment by PeterAlfreadLee** [comment]
Just pushed the code.
I modified `buildDecodingStream` to a parameterized method. Some entries need to be skipped, and the code about skipping them have already been added to `buildDecodingStream`.
Currently the `getInputStream` only called the `buildDecodingStream` followed by `getCurrentStream` , and `currentFolderIndex` & `currentEntryIndex` will be set in `getInputStream`.
The random access using `getInputStream` can work together with sequential access now, and the 
corresponding testcases have already been added to `SevenZFileTest`.

**Comment by PeterAlfreadLee** [comment]
Modified the code a little bit:
```
// if this is called in a random access, then the content methods of previous entry may be null
// the content methods should be set to methods of the first entry as it must not be null
if(currentEntryIndex != entryIndex) {
    int folderFirstFileIndex = archive.streamMap.folderFirstFileIndex[folderIndex];
    SevenZArchiveEntry folderFirstFile = archive.files[folderFirstFileIndex];
    file.setContentMethods(folderFirstFile.getContentMethods());
}
```
When called in a random access, the content methods should be set to methods of the first entry in folder, instead of the methods of the previous entry in folder.
That's because the content methods of the previous entry may be null as it's a random access.

**Comment by bodewig** [comment]
The latest change may only be necessary when the content methods of `file` actually is `null`, which may not be the case when jumping backwards to a position that has already been reached by other means. But an additional check would mainly be an optimization.

Right now I'm not sure the code works in situations where we jump backwards inside the folder or whether we'd need to remove streams from `deferredBlockStreams` otherwise. Probably I'd need to write a test for that :-)

Many thanks!

**Comment by PeterAlfreadLee** [comment]
> The latest change may only be necessary when the content methods of `file` actually is `null`, which may not be the case when jumping backwards to a position that has already been reached by other means. But an additional check would mainly be an optimization.

I agree. I will add the non-null check for the content methods, and the content methods will not be set when the content methods is not null.

> Right now I'm not sure the code works in situations where we jump backwards inside the folder or whether we'd need to remove streams from `deferredBlockStreams` otherwise. Probably I'd need to write a test for that :-)

I agree. I forgot the situations that we may jump backwards inside the folder. It seems I can only reopen the folder and skip the previous files again.

I will modify both of these 2 and push my code today.

**Comment by PeterAlfreadLee** [comment]
Code pushed.
1. The content methods will only be set if it's not null
2. When jumping back to a file that has already been read before, the position of `currentFolderInputStream` will be reset to the beginning of the folder and skip all the files from the beginning of the folder again.
3. The corresponding test cases of jumping back in the same folder have been added.

File: src/main/java/org/apache/commons/compress/archivers/sevenz/SevenZFile.java [hunk 1]
```diff
@@ -1208,6 +1208,42 @@ private InputStream getCurrentStream() throws IOException {
         return deferredBlockStreams.get(0);
     }
 
+    public InputStream getInputStream(SevenZArchiveEntry entry) throws IOException {
```
[the remark this answers is not in the document]
Comment (bodewig) [reply src/main/java/org/apache/commons/compress/archivers/sevenz/SevenZFile.java]: added them myself in 228e7399

File: src/main/java/org/apache/commons/compress/archivers/sevenz/SevenZFile.java [hunk 2]
```diff
@@ -1208,6 +1208,42 @@ private InputStream getCurrentStream() throws IOException {
...
+        int entryIndex = -1;
+        for(int i = 0; i < this.archive.files.length;i++) {
+            if(entry == this.archive.files[i]) {
+                entryIndex = i;
```
[the remark this answers is not in the document]
Comment (bodewig) [reply src/main/java/org/apache/commons/compress/archivers/sevenz/SevenZFile.java]: see 228e7399

**Comment by bodewig** [comment]
Many thanks again @PeterAlfreadLee

**Comment by PeterAlfreadLee** [comment]
Thank you for your reviews @bodewig

**Pull request merged**